File: typescript/oop-patterns/structural-patterns/bridge/bridge.ts
// Абстрактный мост содержит ссылку на конкретную реализацию требуемых операций
// и вызывается клиентом
class AbstractBridge {
  protected implementation: Implementation;

  constructor(implementation: Implementation) {
    this.implementation = implementation;
  }

  public operation(): string {
    const result = this.implementation.operationImplementation();
    return `Абстракция: Базовая операция вместе с конкретной реализацией:\n${result}`;
  }
}

// Абстрактную часть можно расширять независимо от реализаций
class ExtendedAbstractBridge extends AbstractBridge {
  public operation(): string {
    const result = this.implementation.operationImplementation();
    return `Расширенная Абстракция: Расширенная операция вместе с конкретной 
реализацией:\n${result}`;
  }
}

// Интерфейс для каждого класса реализации
interface Implementation {
  operationImplementation(): string;
}

// Конкретная реализация представляет из себя конкретные операции, которые потом
// будут использованы через абстрактный мост клиентом
// таким образом, при условии соответствия интерфейсу реализации, можно
// использовать разные внешние API объекты и т.д. внутри разных реализаций
class ConcreteImplementationA implements Implementation {
  public operationImplementation(): string {
    return 'Конкретная реализация типа "А": Результат на платформе "А".';
  }
}

class ConcreteImplementationB implements Implementation {
  public operationImplementation(): string {
    return 'Конкретная реализация типа "Б": Результат на платформе "Б".';
  }
}

//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
//------------------------------- Клиентский код -------------------------------
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
let implementation = new ConcreteImplementationA();
let abstraction = new AbstractBridge(implementation);
console.log(abstraction.operation());

implementation = new ConcreteImplementationB();
abstraction = new ExtendedAbstractBridge(implementation);
console.log(abstraction.operation());

File: go/oop-patterns/structural-patterns/bridge/bridge.go
package main

import "fmt"

// Абстрактный мост содержит ссылку на конкретную реализацию требуемых операций
// и вызывается клиентом
type abstractBridge struct {
	implementation implementation
}

func (b *abstractBridge) operation() string {
	result := b.implementation.operationImplementation()
	return fmt.Sprintf("Абстракция: Базовая операция вместе с конкретной реализацией:\n%v", result)
}

func newAbstractBridge(i implementation) *abstractBridge {
	return &abstractBridge{implementation: i}
}

// Абстрактную часть можно расширять независимо от реализаций
type extendedAbstractBridge struct {
	abstractBridge
}

func (b *extendedAbstractBridge) operation() string {
	result := b.implementation.operationImplementation()
	return fmt.Sprintf(`Расширенная абстракция: Расширенная операция вместе с 
конкретной реализацией: %v`, result)
}

func newExtendedAbstractBridge(i implementation) *extendedAbstractBridge {
	return &extendedAbstractBridge{abstractBridge: abstractBridge{implementation: i}}
}

// Интерфейс для каждого класса реализации
type implementation interface {
	operationImplementation() string
}

// Конкретная реализация представляет из себя конкретные операции, которые потом
// будут использованы через абстрактный мост клиентом
// таким образом, при условии соответствия интерфейсу реализации, можно
// использовать разные внешние API объекты и т.д. внутри разных реализаций
type concreteImplementationA struct {
}

func (i *concreteImplementationA) operationImplementation() string {
	return "Конкретная реализация типа \"А\": Результат на платформе \"А\"."
}

type concreteImplementationB struct {
}

func (i *concreteImplementationB) operationImplementation() string {
	return "Конкретная реализация типа \"Б\": Результат на платформе \"Б\"."
}

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// ------------------------------- Клиентский код -------------------------------
// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
func main() {
	implA := concreteImplementationA{}
	implB := concreteImplementationB{}
	absBridge := newAbstractBridge(&implA)
	extendedAbsBridge := newExtendedAbstractBridge(&implB)

	fmt.Println(absBridge.operation())
	fmt.Println(extendedAbsBridge.operation())
}
